File: webviewerwidgets/wgnote/wgnote.py
import aiosqlite

async def get_data (queries):
    ret = {}
    try:
        dbpath = queries['dbpath']
        tab = queries['tab']
        rowkey = queries['rowkey']
        note = queries['note'].replace('"', "'")
        conn = await aiosqlite.connect(dbpath)
        c = await conn.cursor()
        if tab == 'variant':
            q = 'update {} set base__note="{}" where base__uid={}'.format(
                tab, note, rowkey)
            await c.execute(q)
        elif tab == 'gene':
            q = 'update {} set base__note="{}" where base__hugo="{}"'.format(
                tab, note, rowkey)
            await c.execute(q)
        await conn.commit()
        ret['status'] = 'success'
    except:
        import traceback
        traceback.print_exc()
        ret['status'] = 'fail'
    await c.close()
    await conn.close()
    return ret

File: webviewerwidgets/wgnote/wgnote.js
widgetGenerators['note'] = {
	'variant': {
		'width': 280, 
		'height': 280, 
		'function': function (div, row, tabName) {
            var colNo = infomgr.getColumnNo(tabName, 'base__note');
            var note = row[colNo];
            var button = getEl('button');
            button.textContent = 'Save';
            button.addEventListener('click', function (evt) {
                var note = evt.target.nextSibling.nextSibling.value;
                $grids[tabName].pqGrid('getData')[selectedRowNos[tabName]][colNo] = note;
                $grids[tabName].pqGrid('refresh');
                $.ajax({
                    url: '/result/runwidget/note',
                    data: {'dbpath': dbPath, 
                           'tab': tabName, 
                           'rowkey': row[0],
                           'note': note,
                          },
                    success: function (response) {
                    }
                });
                console.log(note);
            });
            addEl(div, button);
            addEl(div, getEl('br'));
            var textbox = getEl('textarea');
            textbox.style.width = '100%';
            textbox.style.height = 'calc(100% - 24px)';
            textbox.value = note;
            addEl(div, textbox);
		}
	},
	'gene': {
		'width': 280,
		'height': 280,
		'function': function (div, row, tabName) {
            var colNo = infomgr.getColumnNo(tabName, 'base__note');
            var note = row[colNo];
            var button = getEl('button');
            button.textContent = 'Save';
            button.addEventListener('click', function (evt) {
                var note = evt.target.nextSibling.nextSibling.value;
                $grids[tabName].pqGrid('getData')[selectedRowNos[tabName]][colNo] = note;
                $grids[tabName].pqGrid('refresh');
                $.ajax({
                    url: '/result/runwidget/note',
                    data: {'dbpath': dbPath, 
                           'tab': tabName, 
                           'rowkey': row[0],
                           'note': note,
                          },
                    success: function (response) {
                    }
                });
                console.log(note);
            });
            addEl(div, button);
            addEl(div, getEl('br'));
            var textbox = getEl('textarea');
            textbox.style.width = '100%';
            textbox.style.height = 'calc(100% - 24px)';
            textbox.value = note;
            addEl(div, textbox);
		}
	}
}
